File: packages/silo-client-java/src/main/java/in/org/quicko/silo/client/media/Media.java
package in.org.quicko.silo.client.media;

import com.fasterxml.jackson.databind.JsonNode;
import in.org.quicko.silo.client.RequestOptions;
import in.org.quicko.silo.client.pagination.PageWindow;
import in.org.quicko.silo.client.pagination.RowBatch;
import in.org.quicko.silo.client.pagination.RowLoader;
import in.org.quicko.silo.client.scope.DeleteOptions;
import in.org.quicko.silo.client.transport.ApiPath;
import in.org.quicko.silo.client.transport.JsonValues;
import in.org.quicko.silo.client.transport.Transport;
import in.org.quicko.silo.client.transport.TransportRequest;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * The media library. It is instance-global and never scoped to a project or an
 * environment, which is why it hangs off {@code silo} rather than off a scope.
 */
public final class Media {
  /** Caps a bulk delete before it reaches the server's own cap: a local refusal
   *  reads better than a 400 for the 101st id. */
  private static final int BulkDeleteCap = 100;

  private final Transport transport;
  private final MediaFolders folders;

  public Media(Transport transport) {
    this.transport = transport;
    this.folders = new MediaFolders(transport);
  }

  /** The folder tree, which exists as records rather than as a side effect of
   *  what has been filed. */
  public MediaFolders folders() {
    return folders;
  }

  public MediaAsset upload(MediaUpload input) {
    return upload(input, RequestOptions.none());
  }

  public MediaAsset upload(MediaUpload input, RequestOptions options) {
    JsonNode payload = transport.upload(
        TransportRequest.post(ApiPath.media()).options(options).build(),
        MediaParts.form(input.bytes(), input.filename(), input.contentType(), input.folder()));
    return new MediaAsset(transport, MediaMapper.toRecord(payload));
  }

  public MediaAsset get(String id) {
    return get(id, RequestOptions.none());
  }

  public MediaAsset get(String id, RequestOptions options) {
    JsonNode payload = transport.json(
        TransportRequest.get(ApiPath.mediaAsset(id)).options(options).build());
    return new MediaAsset(transport, MediaMapper.toRecord(payload));
  }

  public MediaPage list() {
    return list(MediaQuery.all(), RequestOptions.none());
  }

  public MediaPage list(MediaQuery query) {
    return list(query, RequestOptions.none());
  }

  public MediaPage list(MediaQuery query, RequestOptions options) {
    PageWindow window = new PageWindow(query.limitOrDefault(), query.offsetOrDefault());
    return MediaPage.load(transport, toWireQuery(query), window, options);
  }

  /** One asset at a time, across every page. */
  public MediaStream all() {
    return all(MediaQuery.all(), RequestOptions.none());
  }

  public MediaStream all(MediaQuery query) {
    return all(query, RequestOptions.none());
  }

  public MediaStream all(MediaQuery query, RequestOptions options) {
    Map<String, Object> wireQuery = toWireQuery(query);
    RowLoader<MediaAsset> loader = window -> {
      MediaPage page = MediaPage.load(transport, wireQuery, window, options);
      return new RowBatch<>(page.files(), new PageWindow(page.limit(), page.offset()));
    };
    return new MediaStream(loader, query.limitOrDefault(), options);
  }

  /** One whole page at a time, across every page. */
  public MediaPageStream pages() {
    return pages(MediaQuery.all(), RequestOptions.none());
  }

  public MediaPageStream pages(MediaQuery query) {
    return pages(query, RequestOptions.none());
  }

  public MediaPageStream pages(MediaQuery query, RequestOptions options) {
    return new MediaPageStream(() -> list(query, options), options);
  }

  public List<String> extensions() {
    return extensions(RequestOptions.none());
  }

  /** Every distinct file extension actually in the library. */
  public List<String> extensions(RequestOptions options) {
    JsonNode body = transport.json(
        TransportRequest.get(ApiPath.mediaExtensions()).options(options).build());
    return JsonValues.strings(body.get("items"));
  }

  public MediaDeleteReport deleteMany(List<String> ids) {
    return deleteMany(ids, DeleteOptions.none());
  }

  /**
   * Always answers a 200: every id's outcome lives in the report rather than in
   * a thrown exception, so a partial success is a value rather than something to
   * infer.
   */
  public MediaDeleteReport deleteMany(List<String> ids, DeleteOptions options) {
    if (ids.size() > BulkDeleteCap) {
      throw new IllegalArgumentException(
          "deleteMany accepts at most " + BulkDeleteCap + " ids per request, got " + ids.size());
    }
    return MediaDeleteReport.fromWire(transport.json(
        TransportRequest.post(ApiPath.mediaBulkDelete())
            .body(Map.of("ids", List.copyOf(ids), "force", options.force()))
            .options(options.request())
            .build()));
  }

  private static Map<String, Object> toWireQuery(MediaQuery query) {
    Map<String, Object> wire = new LinkedHashMap<>();
    if (query.text() != null) wire.put("q", query.text());
    if (query.folder() != null) wire.put("folder", query.folder());
    if (query.recursive() != null) wire.put("recursive", query.recursive());
    if (query.type() != null) wire.put("type", query.type());
    if (query.extension() != null) wire.put("ext", query.extension());
    if (query.tag() != null) wire.put("tag", query.tag());
    if (query.modifiedAfter() != null) wire.put("modified_after", query.modifiedAfter().toString());
    if (query.modifiedBefore() != null) wire.put("modified_before", query.modifiedBefore().toString());
    if (query.sort() != null) wire.put("sort", query.sort());
    return wire;
  }
}

File: packages/silo-client-dart/lib/src/media/media.dart
import '../pagination/page_window.dart';
import '../request_options.dart';
import '../scope/delete_options.dart';
import '../transport/api_path.dart';
import '../transport/json_values.dart';
import '../transport/transport.dart';
import '../transport/transport_request.dart';
import 'media_asset.dart';
import 'media_asset_record.dart';
import 'media_delete_report.dart';
import 'media_folders.dart';
import 'media_page.dart';
import 'media_page_stream.dart';
import 'media_parts.dart';
import 'media_query.dart';
import 'media_stream.dart';
import 'media_upload.dart';

/// The media library: how you find or add an asset. `MediaAsset` is what you
/// do to one.
final class Media {
  Media(this._transport) : folders = MediaFolders(_transport);

  static const _bulkDeleteCap = 100;

  final Transport _transport;
  final MediaFolders folders;

  Future<MediaAsset> upload(MediaUpload input, [RequestOptions options = const RequestOptions()]) async => MediaAsset(
    _transport,
    MediaAssetRecord.fromJson(
      await _transport.upload(
        TransportRequest('POST', ApiPath.media(), options: options),
        MediaParts.file(input.bytes, input.filename, input.contentType),
        {if (input.folder case final folder? when folder.isNotEmpty) 'folder': folder},
      ),
    ),
  );

  Future<MediaAsset> get(String id, [RequestOptions options = const RequestOptions()]) async => MediaAsset(
    _transport,
    MediaAssetRecord.fromJson(await _transport.json(TransportRequest('GET', ApiPath.mediaAsset(id), options: options))),
  );

  Future<MediaPage> list([MediaQuery query = const MediaQuery(), RequestOptions options = const RequestOptions()]) =>
      MediaPage.load(_transport, query.toWire(), PageWindow(query.limit ?? 50, query.offset ?? 0), options);

  MediaStream all([MediaQuery query = const MediaQuery(), RequestOptions options = const RequestOptions()]) {
    final wireQuery = query.toWire();
    return MediaStream(
      (window) async {
        final page = await MediaPage.load(_transport, wireQuery, window, options);
        return (rows: page.files, window: PageWindow(page.limit, page.offset));
      },
      query.limit ?? 50,
      options,
    );
  }

  MediaPageStream pages([MediaQuery query = const MediaQuery(), RequestOptions options = const RequestOptions()]) =>
      MediaPageStream(() => list(query, options), options);

  /// Every distinct extension in the library.
  Future<List<String>> extensions([RequestOptions options = const RequestOptions()]) async => JsonValues.strings(
    (await _transport.json(TransportRequest('GET', ApiPath.mediaExtensions(), options: options)))['items'],
  );

  /// At most 100 ids. Per-id outcomes come back in the report.
  Future<MediaDeleteReport> deleteMany(List<String> ids, [DeleteOptions options = const DeleteOptions()]) async {
    if (ids.length > _bulkDeleteCap) {
      throw ArgumentError.value(ids.length, 'ids', 'deleteMany accepts at most $_bulkDeleteCap ids per request');
    }
    return MediaDeleteReport.fromJson(
      await _transport.json(
        TransportRequest(
          'POST',
          ApiPath.mediaBulkDelete(),
          body: {'ids': ids, 'force': options.force},
          options: options,
        ),
      ),
    );
  }
}
